class_name NavMap
extends Node2D

const SOURCE_ID = 0
const NAV_LAYER = 1

@export var _tilemap: TileMap

var _tiles = {"off": Vector2i(8, 0), "on": Vector2i(9, 0)}
var _blocked_cells: Array[Vector2]


func _ready():
	Events.entity_placed.connect(_on_entity_placed)
	Events.entity_removed.connect(_on_entity_removed)


func set_cell(cellv: Vector2i, value: String) -> void:
	if cellv in _blocked_cells:
		return

	_tilemap.set_cell(NAV_LAYER, cellv, SOURCE_ID, _tiles[value])


func _on_entity_placed(entity: Entity, cellv: Vector2):
	if Library.get_entity_name(entity) == "Wire":
		return

	set_cell(cellv, "off")

	if cellv not in _blocked_cells:
		_blocked_cells.append(cellv)


func _on_entity_removed(entity: Entity, cellv: Vector2):
	if Library.get_entity_name(entity) == "Wire":
		return

	_blocked_cells.erase(cellv)

	set_cell(cellv, "on")
